### ISSUE TITLE
First attempt on using GHSCI on Linux

### ISSUE THREAD

**Issue by benizar** [healthysustainablecities/global-indicators#256]
Dear developers,

Perhaps this issue could be labeled as *dependencies* for Linux.

...

We (Marc, Xavi and I) wanted to test this software on a workstation with a lot more hardware resources than our PCs, but this machine only has a Linux Ubuntu 22.04 installed (docker v23.0.4, with the compose command). The program is containerized so it should work on Linux too. However we found some small issues that we think are platform related. We tested different versions being the last one `global-indicators-4.2.1` and all presented two similar problems in Linux.

## First problem
The `docker-compose.yml` is deployed and pulls the images without problems. However, the first attempt failed because of two *permission denied error*.

The recently added `instructions.sh` doesn't have execution permissions:

```bash
benizar@alda-pro:~/sandbox/global-indicators-4.2.1$ bash ./global-indicators.sh 

Global Healthy and Sustainable City Indicators
4.2.1

https://healthysustainablecities.org (1000 Cities Challenge)
https://global-healthy-liveable-cities.github.io/ (Software guide)

[+] Running 6/6
 ✔ ghsci 5 layers [⣿⣿⣿⣿⣿]      0B/0B      Pulled                                                                                                                                                            109.9s 
   ✔ 26c5c85e47da Pull complete                                                                                                                                                                              13.1s 
   ✔ 4137f808cdc4 Pull complete                                                                                                                                                                              82.1s 
   ✔ 41e8be2f54ba Pull complete                                                                                                                                                                              99.0s 
   ✔ 0509096dcae9 Pull complete                                                                                                                                                                              99.0s 
   ✔ 35c7560b7494 Pull complete                                                                                                                                                                              99.1s 
[+] Running 4/4
 ✔ Network global-indicators-421_default   Created                                                                                                                                                            0.1s 
 ✔ Volume "global-indicators-421_db-data"  Created                                                                                                                                                            0.0s 
 ✔ Container ghscic_postgis                Started                                                                                                                                                            3.9s 
 ✔ Container ghsci                         Started                                                                                                                                                            3.8s 
/bin/bash: line 1: ./instructions.sh: Permission denied
Please ensure that Docker Desktop is installed and running (https://www.docker.com/products/docker-desktop/). Docker Desktop includes Docker Compose, which is required to run this software.
```

This can be solved manually using:

```bash
chmod +x ./process/instructions.sh
```

A temporary patch for Linux could be to add this step in the command specified in the `docke-compose.yml` file (e.g. `command: ["/bin/bash", "-c", "sleep 3 && chmod +x ./process/instructions.sh  && ./instructions.sh && /bin/bash"]`). Maybe, a different and more stable approach could be to **provide separate scripts for each system** (e.g. `global-indicators-linux.sh` or `global-indicators-ios.sh`.

Another permission issue came when scaffolding the template files:

```bash
# python 1_create_project_configuration_files.py example_ES_Las_Palmas_2023
Creating project configuration files, if not already existing in the configuration folder...
Traceback (most recent call last):
  File "/home/ghsci/work/process/1_create_project_configuration_files.py", line 103, in <module>
    shutil.copyfile(path_file, path_file.replace('templates/', ''))
  File "/env/lib/python3.10/shutil.py", line 256, in copyfile
    with open(dst, 'wb') as fdst:
PermissionError: [Errno 13] Permission denied: './configuration/config.yml'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ghsci/work/process/1_create_project_configuration_files.py", line 106, in <module>
    raise Exception(f'An error occurred: {e}')
Exception: An error occurred: [Errno 13] Permission denied: './configuration/config.yml'
```

The problem seems to be related to having different user ids in the container and on the host. Using [this approach](https://stackoverflow.com/a/56844765) we made it work.

So we added this line to the `ghsci` service in the `docker-compose.yml`
```yaml
    user: "${UID}:${GID}"
```

Then, docker compose is called using:
```bash
        # if Docker exists and is runnable, launch the Docker compose image
        # docker compose -f docker-compose.yml up -d
        UID=$(id -u) GID=$(id -g) docker compose -f docker-compose.yml up -d
```

This way the process could continue as expected.

## Second problem

```bash
# python 2_analyse_region.py example_ES_Las_Palmas_2023

Global Healthy Liveable Cities Indicator Study Collaboration, version 4.2.1

Las Palmas de Gran Canaria (example_ES_Las_Palmas_2023)

Output directory:
  process/data/_study_region_outputs/example_ES_Las_Palmas_2023

A dated copy of project and region parameters has been saved as process/data/_study_region_outputs/example_ES_Las_Palmas_2023/_parameters.yml.

Analysis time zone: Australia/Melbourne (to set time zone for where you are, edit config.yml)

Analysis start:	2023-04-21_0322
Create database:                      0%|                              | (0/13)


Processing _00_create_database.py failed: Command 'python _00_create_database.py example_ES_Las_Palmas_2023' returned non-zero exit status 1.
```

The log shows an error when trying to connect to the database, the host name (*gateway.docker.internal*) is not recognized in Linux but apparently on iOS and on Windows it works just fine.

```bash
Global Healthy Liveable Cities Indicator Study Collaboration, version 4.2.1

Las Palmas de Gran Canaria (example_ES_Las_Palmas_2023)

Output directory:
  process/data/_study_region_outputs/example_ES_Las_Palmas_2023

Connecting to default database to action queries.
Traceback (most recent call last):
  File "/home/ghsci/work/process/subprocesses/_00_create_database.py", line 130, in <module>
    main()
  File "/home/ghsci/work/process/subprocesses/_00_create_database.py", line 26, in main
    conn = psycopg2.connect(
  File "/env/lib/python3.10/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: could not translate host name "gateway.docker.internal" to address: Name or service not known
```

This is explained in detail in [this post](https://medium.com/@TimvanBaarsen/how-to-connect-to-the-docker-host-from-inside-a-docker-container-112b4c71bc66) and the author provides a solution that should work by just editing the `docker-compose.yml` file. Using these suggestions the *ghsci* service now works (I guess that only the second `extra_host` is necessary in this case):

The compose now works and looks like this:
```yaml
#  ...
    volumes:
      - ./:/home/ghsci/work
    user: "${UID}:${GID}"
    extra_hosts:
#      - "host.docker.internal:host-gateway" # From the Tim van Baarsen solution
      - "gateway.docker.internal:host-gateway" # Adjusted to the name in this project

#  ... The rest of the dockerfile
```

## Suggestion

It seems that only these three minor edits are necessary to execute the GHSCI on Linux. 

We should try this on different machines to see if:

- the behavior is consistent in other Linux.
- these edits are compatible with the software working on Windows or iOS.

We hope this help other Linux user's while these issues can be addressed in the main branch.

Please, let us know if we can help with any more testing, logs or a *pull request*.

Thank you for this interesting project! ;)

**Comment by carlhiggs** [comment]
Dear  Benito and colleagues, @benizar @marcdmallafre @xavidelclos , thank you for the detailed description of the issue and proposed solutions; much appreciated!

I'll look into this later today; the suggestion of a pull request is a great one - please feel welcome to do that, and I could aim to test this on some windows and Linux computers I have access to, and ask colleagues to test for MacOS.

Platform specific yml files could be a good solution, if we can't work platform detection and corresponding rules if they're required into the shell script itself reliably.

Anyway, I'll let you know how I go - but feel free to do a pull request; that would be useful :)

(Apologies for typos, written on mobile)

**Comment by carlhiggs** [comment]
Hi @benizar ,

Thanks again for the detailed issue write and sharing code for suggested fix -- I've now implemented and tested this in pull request #257 .   I think because you aren't yet in a team within our organization I couldn't select to add you as a reviewer, but would be great if you could try this out when you get a chance (perhaps on a Mac OS computer too if possible?).

The main bit I changed was moving instructions.sh to the parent folder as ../.env and editing it to make the set of instructions to display a text environment variable (this avoids the need to set permissions for the file to be executable; overkill to just echo some text), and UID and GID likewise as text environment variables (but empty strings; which avoids a warning that otherwise displayed).   As per the write up in the pull request, in my tests this works on both Linux (Ubuntu) and Windows systems (including WSL). 

Keen to hear your thoughts! 

Carl

**Comment by carlhiggs** [comment]
This has been addressed in #257, now merged to the main branch -- I will update a new formal release with this fix later
